src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -332,15 +358,74 @@ public void CopyChangedValues(TStructuralType original)
                 }
                 else
                 {
-                    // Recursively patch the subtree.
-                    bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
-                    Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+                    // Recursively patch the subtree.                    
+                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+
+                    Type newType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && newType != originalType)
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
+                    }
 
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType parent, string nestedPropertyName, dynamic originalValue, Type newType, Type originalType, Type declaredType)
+        {
+            //As per OASIS discussion, changing a complex type from 1 derived type to another is allowed if both derived type have a common ancestor and the property
+            //is declared in terms of a common ancestor. The logic below checks for a common ancestor. Create a new object of the derived type in delta request.
+            //And copy the common properties.
+
+            Type newBaseType = newType;
+            HashSet<Type> newBaseTypes = new HashSet<Type>();
+
+            //Iterate till you find the declaring base type and add all that to hashset
+            while (newBaseType != null && newBaseType != declaredType)
+            {
+                newBaseTypes.Add(newBaseType);
+                newBaseType = newBaseType.BaseType;
+            }
+
+            newBaseTypes.Add(declaredType);
+
+            //Here original type is the type for original (T) resource.
+            //We will keep  going to base types and finally will get the Common Basetype for the derived complex types in to the originalType variable.
+            
+            //The new Original type, means the new complex type (T) which will replace the current complex type.
+            dynamic newOriginalNestedResource = null;
```

Comment:
>dynamic newOriginalNestedResource = null; [](http://example.com/codeflow?start=12&length=41)

In iteration 3 you returned originalNestedResource if no common base type was found. In iteration 4 the behavior changed to return null if no common base type was found. Was this intentional, or should newOriginalNestedResource default to originalValue if no common base type is found in order to preserve existing behavior?